subxt/src/rpc/lightclient/mod.rs

```diff
@@ -0,0 +1,241 @@
+    ) -> Self {
+        self.potential_relay_chains = Some(potential_relay_chains.collect());
+        self
+    }
+
+    /// The URL of a trusted node from which the chain spec is fetched.
+    ///
+    /// # Note
+    ///
+    /// Incompatible with [`Self::chain_spec`].
+    #[cfg(feature = "jsonrpsee-ws")]
+    pub fn trusted_url(mut self, url: &'a str) -> Self {
+        self.url = Some(url);
+        self
+    }
+
+    /// Build the light client with specified URL to connect to.
+    /// You must provide the port number in the URL.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if being called outside of `tokio` runtime context.
+    pub async fn build(self) -> Result<LightClient, Error> {
+        let mut chain_spec = match (self.chain_spec, self.url) {
+            (Some(chain_spec), None) => serde_json::from_str(chain_spec).unwrap(),
+            #[cfg(feature = "jsonrpsee-ws")]
+            (None, Some(url)) => fetch_url(url).await?,
+            _ => return Err(Error::LightClient(LightClientError::IncompatibleConfig)),
+        };
+
+        // Set custom bootnodes if provided.
+        if let Some(bootnodes) = self.bootnodes {
+            let bootnodes = bootnodes
+                .into_iter()
+                .map(serde_json::Value::String)
+                .collect();
+            if let serde_json::Value::Object(map) = &mut chain_spec {
+                map.insert("bootNodes".to_string(), serde_json::Value::Array(bootnodes));
+            }
+        }
+
+        let config = smoldot_light::AddChainConfig {
+            specification: &chain_spec.to_string(),
+            json_rpc: smoldot_light::AddChainConfigJsonRpc::Enabled {
+                max_pending_requests: self.max_pending_requests,
+                max_subscriptions: self.max_subscriptions,
+            },
+            potential_relay_chains: self.potential_relay_chains.unwrap_or_default().into_iter(),
+            database_content: self.database_content.unwrap_or_default(),
+            user_data: (),
+        };
+
+        LightClient::new(config)
+    }
+}
+
+/// Fetch the chain spec from the URL.
+#[cfg(feature = "jsonrpsee-ws")]
+async fn fetch_url(url: impl AsRef<str>) -> Result<serde_json::Value, Error> {
+    let url = url
+        .as_ref()
+        .parse::<Uri>()
+        .map_err(|_| Error::LightClient(LightClientError::InvalidUrl))?;
+
+    if url.scheme_str() != Some("ws") && url.scheme_str() != Some("wss") {
+        return Err(Error::LightClient(LightClientError::InvalidScheme));
+    }
+
+    let (sender, receiver) = WsTransportClientBuilder::default()
+        .build(url)
+        .await
+        .map_err(|_| LightClientError::Handshake)?;
+
+    let client = ClientBuilder::default()
+        .request_timeout(core::time::Duration::from_secs(180))
+        .max_notifs_per_subscription(4096)
+        .build_with_tokio(sender, receiver);
+
+    client
+        .request("sync_state_genSyncSpec", rpc_params![true])
```

Comment:
I wonder how we'll get hold of this with Pierre's new API; is there some Runtime API to get hold of it for instance? If an RPC API we should use that to help future us, and if not then we should ask Pierre maybe?